File: torch_vs_torchsharp/pytorch/chess_rl_sim.py
""" Simulating greedy RL chess bot, without the chess """


from abc import ABC, abstractmethod
import random
import collections
import time
from typing import Deque
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from tqdm import tqdm


WHITE = 1
BLACK = -1


def main():
    opponent = RandomAgent()

    # print("cpu:")
    # agent = GreedyChessAgent(device="cpu")
    # agent.train_against(opponent, 100)

    # print()

    print("gpu:")
    agent = GreedyChessAgent(device="cuda")
    agent.train_against(opponent, 2000)


class FakeChessGame:
    def __init__(self):
        self.turn = WHITE
        self._done = False
        self._done_prev = False
        self._winner = None
        self._winner_prev = None
        self._total_halfmoves = 0
        self._state = np.random.rand(8, 8, 8).astype(np.float32)
        self._state_prev = self._state

    def legal_moves(self):
        return [random.randint(0, 4096) for _ in range(random.randint(1, 20))]

    def winner(self):
        return self._winner

    def step(self, move: int):
        self._total_halfmoves += 1
        self._state_prev = self._state
        self._state = np.random.rand(8, 8, 8).astype(np.float32)
        self._done_prev = self._done
        self._done = self._total_halfmoves > 50
        if self._done:
            self._winner_prev = self._winner
            self._winner = random.choice([None, -1, 1])
        return self._done, random.randint(-1, 0)

    def undo(self):
        self._done = self._done_prev
        self._winner = self._winner_prev
        self._total_halfmoves -= 1
        self._state = self._state_prev

    def state_np(self):
        return self._state



class ChessAgent(ABC):
    @abstractmethod
    def get_action(self, game: FakeChessGame) -> int:
        pass


class RandomAgent(ChessAgent):
    def get_action(self, game: FakeChessGame) -> int:
        return random.choice(game.legal_moves())


class ValueNetwork(nn.Module):
    def __init__(self, input_dim=(8, 8, 8)):
        super(ValueNetwork, self).__init__()
        self.conv1 = nn.Conv2d(8, 32, kernel_size=3, padding=1)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        self.conv3 = nn.Conv2d(64, 32, kernel_size=3, padding=1)

        self.fc1 = nn.Linear(32 * 8 * 8, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 1)

        self.dropout = nn.Dropout(0.3)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = F.relu(self.conv3(x))

        x = x.view(x.size(0), -1)
        x = F.relu(self.fc1(x))
        x = self.dropout(x)
        x = F.relu(self.fc2(x))
        x = torch.tanh(self.fc3(x))

        return x


class ExperienceReplay:
    def __init__(self, capacity=128):
        self.buffer: Deque = collections.deque(maxlen=capacity)

    def push(self, state, next_state, reward):
        self.buffer.append((state, next_state, reward))

    def sample(self, batch_size):
        return random.sample(self.buffer, min(len(self.buffer), batch_size))

    def __len__(self):
        return len(self.buffer)


class GreedyChessAgent:
    def __init__(self, player=WHITE, lr=1e-3, gamma=0.99, tau=0.001, batch_size=32, device="cpu"):
        self.player = player
        self.device = torch.device(device)

        # Networks
        self.value_net = ValueNetwork().to(self.device)
        self.target_net = ValueNetwork().to(self.device)
        self.target_net.load_state_dict(self.value_net.state_dict())

        # Training parameters
        self.optimizer = optim.Adam(self.value_net.parameters(), lr=lr)
        self.gamma = gamma
        self.tau = tau
        self.batch_size = batch_size

        # Experience replay
        self.replay_buffer = ExperienceReplay()

        # Gradient clipping
        self.max_grad_norm = 1.0


    def get_action(self, env: FakeChessGame) -> int:
        assert self.player == WHITE
        assert env.turn == self.player
        legal_moves = list(env.legal_moves())

        if not legal_moves:
            raise ValueError("No legal moves available")

        resulting_states = []
        for move in legal_moves:
            env.step(move)
            resulting_states.append(env.state_np())
            env.undo()

        state_tensors = torch.stack(
            [torch.tensor(state, dtype=torch.float32) for state in resulting_states]
        ).to(self.device)

        with torch.no_grad():
            values = self.value_net(state_tensors).squeeze()

        best_idx = int(torch.argmax(values).item())
        return legal_moves[best_idx]

    def _state_to_tensor(self, state_array: np.ndarray):
        return (
            torch.tensor(state_array, dtype=torch.float32).unsqueeze(0).to(self.device)
        )

    def _update_target_network(self):
        for target_param, param in zip(
            self.target_net.parameters(), self.value_net.parameters()
        ):
            target_param.data.copy_(
                self.tau * param.data + (1.0 - self.tau) * target_param.data
            )

    def train_against(
        self,
        opponent: ChessAgent,
        n_episodes: int,
    ):
        pbar = tqdm(range(n_episodes), desc="Training Episodes")
        start = time.time()
        total_steps = 0
        for _ in pbar:
            game = FakeChessGame()
            done = False
            players = {WHITE: self, BLACK: opponent}
            prev_state = game.state_np()
            game_length = 0
            episode_reward = 0.0

            while not done:
                player: ChessAgent = players[game.turn]  # type: ignore
                move = player.get_action(game)
                done, reward = game.step(move)
                total_steps += 1
                state = game.state_np()
                game_length += 1

                if game.turn != self.player:
                    self.add_experience(prev_state, state, reward)
                    episode_reward += reward

                if game.turn == self.player:
                    prev_state = state

                self.train_step()
        end = time.time()
        duration = end - start
        ep_rate = n_episodes / duration
        pos_rate = total_steps / duration
        print(f"Done training in {duration:.2f}s. {ep_rate:.2f} eps/sec, {pos_rate:.2f} moves/sec")

    def train_step(self):
        if len(self.replay_buffer) < self.batch_size:
            return None

        batch = self.replay_buffer.sample(self.batch_size)
        states, next_states, rewards = zip(*batch)

        states_t = torch.stack(
            [torch.tensor(s, dtype=torch.float32) for s in states]
        ).to(self.device)
        next_states_t = torch.stack(
            [torch.tensor(s, dtype=torch.float32) for s in next_states]
        ).to(self.device)
        rewards_t = torch.tensor(rewards, dtype=torch.float32).to(self.device)

        current_values = self.value_net(states_t).squeeze()

        with torch.no_grad():
            next_values = self.target_net(next_states_t).squeeze()
            target_values = rewards_t + self.gamma * next_values

        loss = F.mse_loss(current_values, target_values)

        self.optimizer.zero_grad()
        loss.backward()

        # Gradient clipping
        torch.nn.utils.clip_grad_norm_(self.value_net.parameters(), self.max_grad_norm)

        self.optimizer.step()

        self._update_target_network()

        return loss.item()

    def add_experience(self, state, next_state, reward):
        self.replay_buffer.push(state, next_state, reward)


if __name__ == "__main__":
    main()

File: torch_vs_torchsharp/csharp/ChessRlSim.cs
using System.Diagnostics;
using TorchSharp;
using TorchSharp.Modules;
using static TorchSharp.torch;

namespace csharp;

public static class ChessRlSim
{
    public static void Run(string[] args)
    {
        var device = args.Contains("cpu") ? "cpu" : args.Contains("gpu") ? "gpu" : "both";

        if (device is "cpu" or "both")
        {
            Console.WriteLine("cpu:");
            var agent = new GreedyChessAgent(device: "cpu");
            var opponent = new RandomAgent();
            agent.TrainAgainst(opponent, 100);
        }

        if (device is "gpu" or "both")
        {
            Console.WriteLine();
            Console.WriteLine("gpu:");
            var agent = new GreedyChessAgent(device: "cuda");
            var opponent = new RandomAgent();
            agent.TrainAgainst(opponent, 2000);
        }
    }
}

public class FakeChessGame
{
    public const int White = 1;
    public const int Black = -1;

    public int turn = White;
    private bool _done = false, _donePrev = false;
    private int? _winner = null, _winnerPrev = null;
    private int _totalHalfmoves = 0;
    private float[,,] _state;
    private float[,,] _statePrev;
    private static Random rng = new Random();

    public FakeChessGame()
    {
        _state = RandomState();
        _statePrev = _state;
    }

    public List<int> LegalMoves()
    {
        int n = rng.Next(1, 21);
        var moves = new List<int>();
        for (int i = 0; i < n; i++)
            moves.Add(rng.Next(0, 4097));
        return moves;
    }

    public int? Winner() => _winner;

    public (bool, int) Step(int move)
    {
        _totalHalfmoves++;
        _statePrev = _state;
        _state = RandomState();
        _donePrev = _done;
        _done = _totalHalfmoves > 50;
        if (_done)
        {
            _winnerPrev = _winner;
            int[] choices = { 0, -1, 1 };
            _winner = choices[rng.Next(choices.Length)];
        }
        return (_done, rng.Next(-1, 1));
    }

    public void Undo()
    {
        _done = _donePrev;
        _winner = _winnerPrev;
        _totalHalfmoves--;
        _state = _statePrev;
    }

    public float[,,] AsFloatArray() => _state;

    private float[,,] RandomState()
    {
        var arr = new float[8, 8, 8];
        for (int i = 0; i < 8; i++)
        for (int j = 0; j < 8; j++)
        for (int k = 0; k < 8; k++)
            arr[i, j, k] = (float)rng.NextDouble();
        return arr;
    }
}

public interface IChessAgent
{
    int GetAction(FakeChessGame game);
}

public class RandomAgent : IChessAgent
{
    private static readonly Random Rng = new();

    public int GetAction(FakeChessGame game)
    {
        var moves = game.LegalMoves();
        return moves[Rng.Next(moves.Count)];
    }
}

public sealed class ValueNetwork : nn.Module
{
    private readonly Sequential _seq;

    public ValueNetwork(string name = "ValueNetwork") : base(name)
    {
        _seq = nn.Sequential(
            nn.Conv2d(8, 32, 3, padding: 1),
            nn.ReLU(),
            nn.Conv2d(32, 64, 3, padding: 1),
            nn.ReLU(),
            nn.Conv2d(64, 32, 3, padding: 1),
            nn.ReLU(),
            nn.Flatten(),
            nn.Linear(32 * 8 * 8, 128),
            nn.ReLU(),
            nn.Dropout(0.3),
            nn.Linear(128, 64),
            nn.ReLU(),
            nn.Linear(64, 1),
            nn.Tanh()
        );
        RegisterComponents();
    }

    public Tensor Forward(Tensor x)
    {
        return _seq.forward(x);
    }
}

public class ExperienceReplay
{
    private readonly int _capacity;
    private readonly Queue<(float[,,], float[,,], float)> _buffer;

    public ExperienceReplay(int capacity = 128)
    {
        _capacity = capacity;
        _buffer = new Queue<(float[,,], float[,,], float)>();
    }

    public void Push(float[,,] state, float[,,] nextState, float reward)
    {
        if (_buffer.Count >= _capacity)
            _buffer.Dequeue();
        _buffer.Enqueue((state, nextState, reward));
    }

    public List<(float[,,], float[,,], float)> Sample(int batchSize)
    {
        var list = _buffer.ToList();
        var rng = new Random();
        return list.OrderBy(_ => rng.Next()).Take(Math.Min(batchSize, list.Count)).ToList();
    }

    public int Count => _buffer.Count;
}

public class GreedyChessAgent : IChessAgent
{
    private readonly int _player;
    private readonly ValueNetwork _valueNet;
    private readonly ValueNetwork _targetNet;
    private readonly optim.Optimizer _optimizer;
    private readonly float _gamma;
    private readonly float _tau;
    private readonly int _batchSize;
    private readonly ExperienceReplay _replayBuffer;
    private const float MaxGradNorm = 1.0f;
    private readonly Device _device;

    public GreedyChessAgent(
        int player = FakeChessGame.White,
        float lr = 1e-3f,
        float gamma = 0.99f,
        float tau = 0.001f,
        int batchSize = 32,
        string device = "cpu"
    )
    {
        _player = player;
        _device = torch.device(device);
        _valueNet = new ValueNetwork().to(_device);
        _targetNet = new ValueNetwork().to(_device);
        _targetNet.load_state_dict(_valueNet.state_dict());
        _optimizer = optim.Adam(_valueNet.parameters(), lr: lr);
        _gamma = gamma;
        _tau = tau;
        _batchSize = batchSize;
        _replayBuffer = new ExperienceReplay();
    }

    public int GetAction(FakeChessGame env)
    {
        if (_player != FakeChessGame.White) throw new Exception("Only WHITE supported");
        if (env.turn != _player) throw new Exception("Not this agent's turn");
        var legalMoves = env.LegalMoves();
        if (legalMoves.Count == 0) throw new Exception("No legal moves available");
        var resultingStates = new List<float[,,]>();
        foreach (var move in legalMoves)
        {
            env.Step(move);
            resultingStates.Add(env.AsFloatArray());
            env.Undo();
        }

        var stateTensors = stack(resultingStates
                .Select(arr => tensor(arr, dtype: ScalarType.Float32))
                .ToArray())
            .to(_device);
        using (no_grad())
        {
            var values = _valueNet.Forward(stateTensors).squeeze();
            var bestIdx = values.argmax().ToInt32();
            return legalMoves[bestIdx];
        }
    }

    private void UpdateTargetNetwork()
    {
        var targetParams = _targetNet.parameters().ToArray();
        var valueParams = _valueNet.parameters().ToArray();
        for (int i = 0; i < targetParams.Length; i++)
        {
            targetParams[i].mul_(1 - _tau).add_(valueParams[i].mul(_tau));
        }
    }

    public void TrainAgainst(IChessAgent opponent, int nEpisodes)
    {
        var sw = Stopwatch.StartNew();
        var totalSteps = 0;
        var numEpsPerDispose = 100;

        var d = NewDisposeScope();

        for (var ep = 0; ep < nEpisodes; ep++)
        {
            if (ep > 0 && ep % numEpsPerDispose == 0)
            {
                d.Dispose();
                d = NewDisposeScope();
            }

            var game = new FakeChessGame();
            var done = false;
            var players = new Dictionary<int, IChessAgent>
            {
                { FakeChessGame.White, this },
                { FakeChessGame.Black, opponent }
            };
            var prevState = game.AsFloatArray();

            while (!done)
            {
                var playerAgent = players[game.turn];
                var move = playerAgent.GetAction(game);
                (done, var reward) = game.Step(move);
                totalSteps++;
                var state = game.AsFloatArray();
                if (game.turn != _player)
                {
                    AddExperience(prevState, state, reward);
                }
                if (game.turn == _player)
                    prevState = state;
                TrainStep();
            }
        }
        sw.Stop();
        var duration = sw.Elapsed.TotalSeconds;
        var epRate = nEpisodes / duration;
        var posRate = totalSteps / duration;
        Console.WriteLine($"Done training in {duration:F2}s. {epRate:F2} eps/sec, {posRate:F2} moves/sec");
    }

    private void TrainStep()
    {
        if (_replayBuffer.Count < _batchSize) return;
        var batch = _replayBuffer.Sample(_batchSize);
        var states = batch.Select(x => tensor(x.Item1, dtype: ScalarType.Float32)).ToArray();
        var nextStates = batch.Select(x => tensor(x.Item2, dtype: ScalarType.Float32)).ToArray();
        var rewards = batch.Select(x => x.Item3).ToArray();
        var statesT = stack(states).to(_device);
        var nextStatesT = stack(nextStates).to(_device);
        var rewardsT = tensor(rewards, dtype: ScalarType.Float32).to(_device);
        var currentValues = _valueNet.Forward(statesT).squeeze();
        using (no_grad())
        {
            var nextValues = _targetNet.Forward(nextStatesT).squeeze();
            var targetValues = rewardsT + _gamma * nextValues;
            var loss = nn.functional.mse_loss(currentValues, targetValues);
            _optimizer.zero_grad();
            loss.backward();
            nn.utils.clip_grad_norm_(_valueNet.parameters(), MaxGradNorm);
            _optimizer.step();
            UpdateTargetNetwork();
        }
    }

    private void AddExperience(float[,,] state, float[,,] nextState, float reward)
    {
        _replayBuffer.Push(state, nextState, reward);
    }
}
